### ISSUE TITLE
Hello, I'm curious about how you implemented the paint by example script?

### ISSUE THREAD

**Issue by thebestYezhang** [buxiangzhiren/Asymmetric_VQGAN#6]
Firstly, thank you so much for your open source code. It has to be said that this is an excellent piece of work. I recently ran the code for Paint By Example and did experience the issue of missing details described in your paper. I see in your paper that there is an experimental comparison with Paint By Example. However, the implementation of this script is not found in your Code base. So, can you open up the script for implementing Paint By Example in your article? Or could you send it to me for learning? Thanks again！！！

**Comment by KenanHuang** [comment]
I think you just need to replace the decoder of the Autoencoder_KL with the larger decoder implemented in Asy-VAGAN.
You can follow the instruction of the "inpaint_st.py" and understand what dose it do in the script.
Well as the description in the bottom of this issue #3

**Comment by phani1223** [comment]
Hi, How to implement this? Will your large decoder work for this task? Also, how do we give the reference image as input? Can you provide a sample it's not clear from the issue mentioned here. Thank you!

**Comment by buxiangzhiren** [comment]
Replace this line (https://github.com/Fantasy-Studio/Paint-by-Example/blob/main/scripts/inference.py#L346) in paint by example repo with our line (https://github.com/buxiangzhiren/Asymmetric_VQGAN/blob/main/inpaint_st.py#L149)

**Comment by buxiangzhiren** [comment]
> Hi, How to implement this? Will your large decoder work for this task? Also, how do we give the reference image as input? Can you provide a sample it's not clear from the issue mentioned here. Thank you!

We use the original paint by example model to generate results and replace their decoder with our base/larger decoder to get the final images. Hence, you just need to give the reference image to the original paint by example model.

**Comment by phani1223** [comment]
I got it. Thanks for replying. Also, I need your expertise in my use case : Transfering a person from a reference image to the initial image. do you think modifying this with paint by example with your decoder works can modify the image with the below features

1. Vibe - ( changing the position of the person based on the vibe of initial image)
eg : The generated image should make the person fly in air if they are sky diving or jumping on ice
2. Mode - ( changing the angle and position of the person based on the type of mode of initial image)
eg: if they are taking a selfie or a wide angle image. should position accordingly
3. Position of objects in the initial image - ( should change the position of the person based on the objects as reference in the initial image )
eg: if a person in the initial images turns back, the generated image also should have the person in reference image turn back )

I just want to know if paint by example with your decoder can generalize such scenarios.  ( pov: I am training this on 30K images of people category from COCO ). Thanks a lot!

**Comment by buxiangzhiren** [comment]
> I got it. Thanks for replying. Also, I need your expertise in my use case : Transfering a person from a reference image to the initial image. do you think modifying this with paint by example with your decoder works can modify the image with the below features
> 
> 1. Vibe - ( changing the position of the person based on the vibe of initial image)
>    eg : The generated image should make the person fly in air if they are sky diving or jumping on ice
> 2. Mode - ( changing the angle and position of the person based on the type of mode of initial image)
>    eg: if they are taking a selfie or a wide angle image. should position accordingly
> 3. Position of objects in the initial image - ( should change the position of the person based on the objects as reference in the initial image )
>    eg: if a person in the initial images turns back, the generated image also should have the person in reference image turn back )
> 
> I just want to know if paint by example with your decoder can generalize such scenarios. ( pov: I am training this on 30K images of people category from COCO ). Thanks a lot!

Using our decoder combined with “paint by example”  simply cannot address the three scenarios mentioned. This is because, when used directly, the generated parts in the image are entirely determined by “paint by example” itself. It lacks the necessary guidance for transformations such as turning or adjusting the camera angle.

However, if we combine our decoder with “paint by example” and explicitly design some guidance mechanisms, improvements can be achieved. For example, leveraging the conditional branch of our decoder to extract features from the background regions and using these features to generate semantic guidance, such as pose or camera view, can help guide the denoising process in “paint by example.”

**Comment by phani1223** [comment]
> Replace this line (https://github.com/Fantasy-Studio/Paint-by-Example/blob/main/scripts/inference.py#L346) in paint by example repo with our line (https://github.com/buxiangzhiren/Asymmetric_VQGAN/blob/main/inpaint_st.py#L149)

Hi @buxiangzhiren , I tried making the following changes. 

1. loaded the first_stage_model and applied the decode function which looks like this 
````
config1 = OmegaConf.load('autoencoder_kl_32x32x4.yaml')
 first_stage_model = load_model_from_config(config1,'base.ckpt')
 first_stage_model = first_stage_model.to(device)
 x_samples_ddim = first_stage_model.decode(1. / 0.18215 * samples_ddim, image_tensor, mask_tensor)
````

2. Changed the paint by example's AutoencoderKL class to AutoencoderKL class from your repo as your decoder function takes mask also in autoencoder.py
3. Imported the decoder from my_decoder.py

When I run with a sample image I get the same original image as the output 

Can you confirm if these are the exact changes I should make?  Thanks In advance.

**Comment by buxiangzhiren** [comment]
> > Replace this line (https://github.com/Fantasy-Studio/Paint-by-Example/blob/main/scripts/inference.py#L346) in paint by example repo with our line (https://github.com/buxiangzhiren/Asymmetric_VQGAN/blob/main/inpaint_st.py#L149)
> 
> Hi @buxiangzhiren , I tried making the following changes.
> 
> 1. loaded the first_stage_model and applied the decode function which looks like this
> 
> ```
> config1 = OmegaConf.load('autoencoder_kl_32x32x4.yaml')
>  first_stage_model = load_model_from_config(config1,'base.ckpt')
>  first_stage_model = first_stage_model.to(device)
>  x_samples_ddim = first_stage_model.decode(1. / 0.18215 * samples_ddim, image_tensor, mask_tensor)
> ```
> 
> 2. Changed the paint by example's AutoencoderKL class to AutoencoderKL class from your repo as your decoder function takes mask also in autoencoder.py
> 3. Imported the decoder from my_decoder.py
> 
> When I run with a sample image I get the same original image as the output
> 
> Can you confirm if these are the exact changes I should make? Thanks In advance.

Your code appears to be correct. Please verify that your mask tensor and image tensor are in the correct format. Moreover, you can try one more sample to see what will happen.

**Comment by phaniratan1234** [comment]
For anyone who still didn't figure this out, check the steps below

Step 1: Clone the paint by example repo.

Step 2: Create a virtual environment and install the dependencies mentioned in their config. Install only the specific version mentioned in environment.yaml, or you will be in dependency hell.

Step 3: Try running their inference script and see if you are able to replicate the inference setup. Check the readme, it has clear instructions.

Step 4: Add the asymmetric GAN's decoder yaml file to your configs/ - you can choose any of the asymmetric GAN decoders they have, and also download the checkpoint of the decoder into the checkpoints/ folder

Step 5: Replace the code line from inference.py in your repo with the one shared below

Replace this line 
`x_samples_ddim = model.decode_first_stage(samples_ddim)`
 
to this

```
 config1 = OmegaConf.load('configs/autoencoder_large.yaml') # change this to the decoder you want
   first_stage_model = load_model_from_config(config1,'checkpoints/larger1.5.ckpt')  # load the checkpoint you want
   first_stage_model = first_stage_model.to(device)
   print("loaded checkpoint and config!")
   samples_ddim = samples_ddim.half()    # only if you are running on CPU or MPS
    image_tensor = image_tensor.half()     # only if you are running on CPU or MPS
    mask_tensor = mask_tensor.half()      # only if you are running on CPU or MPS
    samples_ddim = samples_ddim.to(device)        # the model is loaded on device so to avoid error we send to device
    image_tensor = image_tensor.to(device)         # the model is loaded on device so to avoid error we send to device
    mask_tensor = mask_tensor.to(device)          # the model is loaded on device so to avoid error we send to device
    
    x_samples_ddim = first_stage_model.decode(1. / 0.18215 * samples_ddim,image_tensor, mask_tensor)
```


Step 6: When you call **first_stage_model.decode** it searches for the definition in autoencoder.py. How does it know?
Well, it is mentioned in your yaml file like this **target: ldm.models.autoencoder.AutoencoderKL**. So, your yaml file is important, for different decoders we have different yaml files. So place the correct one

The AutoencoderKL class has a requirement of this line **from ldm.modules.diffusionmodules.my_decoder_large import Decoder**

So now copy the my_decoder_large.py from asymmetric GAN's repo from ldm/modules/diffusionmodules/my_decoder_large.py to the same path in your repo : ldm/modules/diffusionmodules/

so now the my_decoder_large.py has a requirement of the folder saicinpainting/ from asymmetric VQGAN so directly copy the entire folder and keep it in your repo

That's it, you are good to go, run the inference now and check if you see any difference.

Step 7: I personally feel that if the reference alone is not able to capture the masked area, the asymmetric GAN's decoder will inpaint the mask from the context of the image tensor passed, not the reference. I don't know why this happens. To stop this, I have used a diffusion bias parameter in the decoder class. The more you keep it, the better it forces the reference image into the masked area. So you can replace the decoder in class in my_decoder_large.py with the one below

```
class Decoder(nn.Module):
    def __init__(self, *, ch, out_ch, ch_mult=(1,2,4,8), num_res_blocks,
                 attn_resolutions, dropout=0.0, resamp_with_conv=True, in_channels,
                 resolution, z_channels, give_pre_end=False, tanh_out=False, use_linear_attn=False,
                 attn_type="vanilla", diffusion_bias=0.8, **ignorekwargs):
        super().__init__()
        if use_linear_attn: attn_type = "linear"
        ch = 192
        self.ch = ch
        self.temb_ch = 0
        self.num_resolutions = len(ch_mult)
        self.num_res_blocks = 3
        self.resolution = resolution
        self.in_channels = in_channels
        self.give_pre_end = give_pre_end
        self.tanh_out = tanh_out
        # Add diffusion bias parameter to control blending
        self.diffusion_bias = diffusion_bias

        # compute in_ch_mult, block_in and curr_res at lowest res
        in_ch_mult = (1,)+tuple(ch_mult)
        block_in = ch*ch_mult[self.num_resolutions-1]
        curr_res = resolution // 2**(self.num_resolutions-1)
        self.z_shape = (1,z_channels,curr_res,curr_res)
        print("Working with z of shape {} = {} dimensions.".format(
            self.z_shape, np.prod(self.z_shape)))

        # z to block_in
        self.conv_in = torch.nn.Conv2d(z_channels,
                                       block_in,
                                       kernel_size=3,
                                       stride=1,
                                       padding=1)

        # middle
        self.mid = nn.Module()
        self.mid.block_1 = ResnetBlock(in_channels=block_in,
                                       out_channels=block_in,
                                       temb_channels=self.temb_ch,
                                       dropout=dropout)
        self.mid.attn_1 = make_attn(block_in, attn_type=attn_type)
        self.mid.block_2 = ResnetBlock(in_channels=block_in,
                                       out_channels=block_in,
                                       temb_channels=self.temb_ch,
                                       dropout=dropout)

        up_layers = []
        out_ch_ = 192
        res_ch = 768
        stride = 16
        upsample_type = 'deconv'
        while stride > 1:
            stride = stride // 2
            in_ch_ = out_ch_ * 2
            if out_ch_ > res_ch:
                out_ch_ = res_ch
            if stride == 1:
                in_ch_ = res_ch
            layers_ = []
            layers_.append(UpSample(in_ch_, out_ch_, activate_before='relu', activate_after='none', upsample_type=upsample_type))
            up_layers = layers_ + up_layers
            out_ch_ *= 2
        
        self.up_layers = nn.Sequential(*up_layers)
        self.encoder = Conditionencoder(
            in_ch=out_ch,
            up_layers=self.up_layers,
            downsample_layer="conv"
        )
        # upsampling
        self.up = nn.ModuleList()
        for i_level in reversed(range(self.num_resolutions)):
            block = nn.ModuleList()
            attn = nn.ModuleList()
            block_out = ch*ch_mult[i_level]
            for i_block in range(self.num_res_blocks+1):
                block.append(ResnetBlock(in_channels=block_in,
                                         out_channels=block_out,
                                         temb_channels=self.temb_ch,
                                         dropout=dropout))
                block_in = block_out
                if curr_res in attn_resolutions:
                    attn.append(make_attn(block_in, attn_type=attn_type))
            up = nn.Module()
            up.block = block
            up.attn = attn
            if i_level != 0:
                up.upsample = Upsample(block_in, resamp_with_conv)
                curr_res = curr_res * 2
            self.up.insert(0, up) # prepend to get consistent order

        # end
        self.norm_out = Normalize(block_in)
        self.conv_out = torch.nn.Conv2d(block_in,
                                        out_ch,
                                        kernel_size=3,
                                        stride=1,
                                        padding=1)

    def forward(self, z, image, mask):
        self.last_z_shape = z.shape

        # timestep embedding
        temb = None

        # z to block_in
        h = self.conv_in(z)

        # middle
        h = self.mid.block_1(h, temb)
        h = self.mid.attn_1(h)
        h = self.mid.block_2(h, temb)

        masked_image = (1 - mask) * image
        im_x = self.encoder(masked_image, mask)
        
        # upsampling
        for i_level in reversed(range(self.num_resolutions)):
            h_ = im_x[str(tuple(h.shape))]
            mask_ = F.interpolate(mask, size=h.shape[-2:], mode='nearest')
            
            # Apply bias to mask - gives more weight to diffusion output
            biased_mask_ = torch.clamp(mask_ + (1 - mask_) * self.diffusion_bias, 0, 1)
            
            # Use biased mask for blending
            h = h * biased_mask_ + h_ * (1 - biased_mask_)
            
            for i_block in range(self.num_res_blocks+1):
                h = self.up[i_level].block[i_block](h, temb)
                if len(self.up[i_level].attn) > 0:
                    h = self.up[i_level].attn[i_block](h)
            if i_level != 0:
                h = self.up[i_level].upsample(h)
                
        # Final blending with biased mask
        biased_mask = torch.clamp(mask + (1 - mask) * self.diffusion_bias, 0, 1)
        h = h * biased_mask + im_x[str(tuple(h.shape))] * (1 - biased_mask)
        
        # end
        if self.give_pre_end:
            return h

        h = self.norm_out(h)
        h = nonlinearity(h)
        h = self.conv_out(h)
        if self.tanh_out:
            h = torch.tanh(h)
        return h
```

I don't think you would have any issues other than these, so happy to help if you are stuck anywhere. Hope this helps!
